cache/client.go

```diff
@@ -0,0 +1,122 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/merkletree"
+)
+
+func verifyBlockHeader(header types.BlockHeader, minTimestamp types.Timestamp) error {
+	// Check that the timestamp is not too far in the past to be acceptable.
+	if minTimestamp > header.Timestamp {
+		return fmt.Errorf("Block header validation failed: EarlyTimestamp")
+	}
+
+	// Check if the block is in the extreme future. We make a distinction between
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if header.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return fmt.Errorf("Block header validation failed: ExtremeFutureTimestamp")
+	}
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	// This is the last check because it's an expensive check, and not worth
+	// performing if the earlier checks failed.
+	if header.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
+		return fmt.Errorf("Block header validation failed: FutureTimestamp")
+	}
+	return nil
+}
+
+// minimumValidChildTimestamp returns the earliest timestamp that a child node
+// can have while still being valid.
+func minimumValidChildTimestamp(headers []types.BlockHeader, headerIndex int) (minTimestamp types.Timestamp, err error) {
+	// Get the previous MedianTimestampWindow timestamps.
+	windowTimes := make(types.TimestampSlice, types.MedianTimestampWindow)
+	windowTimes[0] = headers[headerIndex].Timestamp
+	parent := headers[headerIndex].ParentID
+	for i := 1; i < int(types.MedianTimestampWindow); i++ {
+		// If the genesis block is 'parent', use the genesis block timestamp
+		// for all remaining times.
+		if parent == (types.BlockID{}) {
+			windowTimes[i] = windowTimes[i-1]
+			continue
+		}
+
+		if (headerIndex - i) < 0 {
+			return 0, fmt.Errorf(
+				"minimumValidChildTimestamp: headers are not sorted properly or 1st header is not genesis header",
+			)
+		}
+		parent = headers[headerIndex-i].ParentID
+		windowTimes[i] = headers[headerIndex-i].Timestamp
+	}
+	sort.Sort(windowTimes)
+
+	// Return the median of the sorted timestamps.
+	return windowTimes[len(windowTimes)/2], nil
+}
+
+func getHeadersSlice(headers []byte) (headersSlice []types.BlockHeader, err error) {
+	headersSlice = make([]types.BlockHeader, len(headers))
+	headersSlice[0] = types.BlockHeader{
+		Timestamp:  types.GenesisTimestamp,
+		MerkleRoot: types.GenesisBlock.MerkleRoot(),
+	}
+	for i := 1; i < len(headers); i++ {
+		headersSlice[i] = types.BlockHeader{
```

Comment:
Can you first create a sub-slice of just this header (a slice of 48 bytes)? I think it will simplify the further extractions.

Reply:
Done.

Reply:
By the way, is the layout here the same like in Sia's `BlockHeader` (because this is how I consider it here)?
`0-8` bytes: Nonce
`8-16` bytes: Timestamp
`16-48` bytes: MerkleRoot

Reply:
[Yes](https://github.com/starius/sialite/blob/64ebbbf3f0e534452ad8adaa6c3febbd40c0e773/cache/build.go#L38) and it is intended.